src/core/lender/tranche/restricted_operator.sol

```diff
@@ -0,0 +1,51 @@
+// Copyright (C) 2020 Centrifuge
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.4.24;
+
+import "./operator.sol";
+
+// RestrictedOperator restricts the allowance of users
+contract RestrictedOperator is Operator {
+    mapping (address => uint) maxCurrency;  // uint(-1) unlimited access by convention
+    mapping (address => uint) maxToken;     // uint(-1) unlimited access by convention
+
+    constructor(address tranche_, address assessor_)
+    Operator(tranche_, assessor_) public {}
+
+    function approve(address usr, uint maxToken_, uint maxCurrency_) public auth {
+        if(investors[usr] == 0) {
+            investors[usr] = 1;
+        }
+        maxCurrency[msg.sender] = maxCurrency_;
+        maxToken[msg.sender] = maxToken_;
+    }
+
+    function supply(uint currencyAmount) public auth_investor {
+        if (maxCurrency[msg.sender] != uint(-1)) {
```

Comment:
Our erc20 contract implements this function quite nicely. I would copy the logic from there.
```
 57     function transferFrom(address src, address dst, uint wad)                                                                                                                     
 58         public returns (bool)                                                                                                                                                     
 59     {                                                                                                                                                                             
 60         require(balanceOf[src] >= wad, "cent/insufficient-balance");                                                                                                              
 61         if (src != msg.sender && allowance[src][msg.sender] != uint(-1)) {                                                                                                        
 62             require(allowance[src][msg.sender] >= wad, "cent/insufficient-allowance");                                                                                            
 63             allowance[src][msg.sender] = sub(allowance[src][msg.sender], wad);                                                                                                    
 64         }                                                                                                                                                                         
 65         balanceOf[src] = sub(balanceOf[src], wad);                                                                                                                                
 66         balanceOf[dst] = add(balanceOf[dst], wad);                                                                                                                                
 67         emit Transfer(src, dst, wad);                                                                                                                                             
 68         return true;                                                                                                                                                              
 69     }                    
```

This translates to:

```
function supply(uint currencyAmount) public auth_investor { 
if (investAllowance[msg.sender] != uint(-1)) {
    investAllowance[msg.sender] = sub(investAllowance[msg.sender], currencyAmount); 
}
super.invest(currencyAmount);
}
```

Reply:
the erc20 is using `sub` and the `require` which is not necessary.

I think require increases the code readability but both ways are okay we could also only use `sub`